src/app-layer-htp.c

```diff
@@ -1810,16 +1854,27 @@ static int HTPCallbackResponseBodyData(htp_tx_data_t *d)
     SCLogDebug("tx_ud->response_body.content_len_so_far %"PRIu64, tx_ud->response_body.content_len_so_far);
     SCLogDebug("hstate->cfg->response.body_limit %u", hstate->cfg->response.body_limit);
 
-    /* within limits, add the body chunk to the state. */
-    if (hstate->cfg->response.body_limit == 0 || tx_ud->response_body.content_len_so_far < hstate->cfg->response.body_limit)
+    /* within limits, add the body chunk to the state.
+       if HTP_STREAM_DEPTH_SET is set, a file is being stored so stream depth is followed,
+       otherwise content length is checked against response body limit. */
+    if ((!(tx_ud->tcflags & HTP_STREAM_DEPTH_SET) &&
+        (hstate->cfg->response.body_limit == 0 || tx_ud->response_body.content_len_so_far < hstate->cfg->response.body_limit)) ||
+        AppLayerHtpCheckStreamDepth(tx_ud))
+
     {
         uint32_t len = (uint32_t)d->len;
+        uint64_t stream_depth = FileReassemblyDepth();
 
-        if (hstate->cfg->response.body_limit > 0 &&
+        if (!(tx_ud->tcflags & HTP_STREAM_DEPTH_SET) && hstate->cfg->response.body_limit > 0 &&
                 (tx_ud->response_body.content_len_so_far + len) > hstate->cfg->response.body_limit)
         {
             len = hstate->cfg->response.body_limit - tx_ud->response_body.content_len_so_far;
             BUG_ON(len > (uint32_t)d->len);
+        } else if ((tx_ud->tcflags & HTP_STREAM_DEPTH_SET) && stream_depth > 0 &&
+                   (tx_ud->response_body.content_len_so_far + len) > stream_depth)
+        {
+            len = stream_depth - tx_ud->response_body.content_len_so_far;
```

Comment:
can this underflow? If stream_depth = 2000, len = 1000, tx_ud->response_body.content_len_so_far = 2500 you get 3500 > 2000, but 2000 - 2500 = -500. Or maybe tx_ud->response_body.content_len_so_far can't go beyond stream_depth?

I think this whole logic should be implemented in a cleaner way, probably in a util func so it's easier to follow.